media-libs/opensubdiv/opensubdiv-3.4.3-r1.ebuild

```diff
@@ -0,0 +1,111 @@
+# Copyright 1999-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+CMAKE_MAKEFILE_GENERATOR=emake
+
+inherit cmake cuda toolchain-funcs
+
+MY_PV="$(ver_rs "1-3" '_')"
+DESCRIPTION="An Open-Source subdivision surface library"
+HOMEPAGE="https://graphics.pixar.com/opensubdiv/docs/intro.html"
+SRC_URI="https://github.com/PixarAnimationStudios/OpenSubdiv/archive/v${MY_PV}.tar.gz -> ${P}.tar.gz"
+
+# Modfied Apache-2.0 license, where section 6 has been replaced.
+# See for example CMakeLists.txt for details.
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="cuda examples opencl openmp ptex tbb test tutorials"
+
+PATCHES=(
+	"${FILESDIR}/${PN}-3.3.0-use-gnuinstalldirs.patch"
+	"${FILESDIR}/${P}-install-tutorials-into-bin.patch"
+)
+
+RESTRICT="!test? ( test )"
+
+RDEPEND="
+	${PYTHON_DEPS}
+	media-libs/glew:=
+	media-libs/glfw:=
+	x11-libs/libXinerama
+	cuda? ( dev-util/nvidia-cuda-toolkit:* )
+	opencl? ( virtual/opencl )
+	ptex? ( media-libs/ptex )
+"
+DEPEND="
+	${RDEPEND}
+	tbb? ( dev-cpp/tbb )
+"
+S="${WORKDIR}/OpenSubdiv-${MY_PV}"
+
+pkg_pretend() {
+	[[ ${MERGE_TYPE} != binary ]] && use openmp && tc-check-openmp
+}
+
+pkg_setup() {
+	[[ ${MERGE_TYPE} != binary ]] && use openmp && tc-check-openmp
+}
+
+src_prepare() {
+	cmake_src_prepare
+	use cuda && cuda_add_sandbox
+}
+
+src_configure() {
+	# GLTESTS are disabled as portage is unable to open a display during test phase
+	local mycmakeargs=(
+		-DGLEW_LOCATION="${EPREFIX}/usr/$(get_libdir)"
+		-DGLFW_LOCATION="${EPREFIX}/usr/$(get_libdir)"
+		-DNO_CLEW=ON
+		-DNO_CUDA=$(usex !cuda)
+		-DNO_DOC=ON
+		-DNO_EXAMPLES=$(usex !examples)
+		-DNO_GLTESTS=ON
+		-DNO_OMP=$(usex !openmp)
+		-DNO_OPENCL=$(usex !opencl)
+		-DNO_PTEX=$(usex !ptex)
+		-DNO_REGRESSION=$(usex !test)
+		-DNO_TBB=$(usex !tbb)
+		-DNO_TESTS=$(usex !test)
+		-DNO_TUTORIALS=$(usex !tutorials)
+	)
+
+	if use cuda; then
+		if [[ *$(gcc-version)* != $(cuda-config -s) ]]; then
```

Comment:
I find the following code works
'''
adrian@office ~ $ if [[ *10.2* != "4.7 4.8 4.9 5.3 5.4 6.3 6.4 7.2 7.3 8.2 8.3 8.4 9.2 9.3 10.2" ]]; then echo "PrintWarning"; fi
PrintWarning
'''
So I wonder what gcc-version is returning

Reply:
This is from the cuda toolkit ebuild:
```
pkg_postinst_check() {
	local a="$(${EROOT}/opt/cuda/bin/cuda-config -s)"
	local b="0.0"
	local v
	for v in ${a}; do
		ver_test "${v}" -gt "${b}" && b="${v}"
	done

	# if gcc and if not gcc-version is at least greatest supported
	if tc-is-gcc && \
		ver_test $(gcc-version) -gt "${b}"; then
			ewarn
			ewarn "gcc > ${b} will not work with CUDA"
			ewarn "Make sure you set an earlier version of gcc with gcc-config"
			ewarn "or append --compiler-bindir= pointing to a gcc bindir like"
			ewarn "--compiler-bindir=${EPREFIX}/usr/*pc-linux-gnu/gcc-bin/gcc${b}"
			ewarn "to the nvcc compiler flags"
			ewarn
	fi
}
```
Again, I don't see a point of making these checks in opensubdiv.

Reply:
> I find the following code works
> 
> ```
> adrian@office ~ $ if [[ *10.2* != "4.7 4.8 4.9 5.3 5.4 6.3 6.4 7.2 7.3 8.2 8.3 8.4 9.2 9.3 10.2" ]]; then echo "PrintWarning"; fi
> PrintWarning
> ```

I think this is not working the way you'd like it to work. If I understand it right, you expect that it *does not* print `PrintWarning`, because 10.2 is in the list.

Reply:
@khumarahn that's how I imagined it, too. @redchillipadi gcc-version returns 10.2 correctly as seen in the log I pasted. So it just seems the logic is inverted currently? :)

Reply:
> `if [[ $(cuda-config -s) != *$(gcc-version)* ]]; then`

I don't quite like this because ` 19.2 == *9.2*`, even though this is unlikely to cause problems. If you really want to give this warning, you could make the check as in the cuda toolkit ebuild... Or just give an unconditional warning of the type "you are enabling cuda, so make sure that your gcc/clang version is supported by the cuda toolkit".

I found that the cuda error messages are quite clear when the compiler version is not supported, and that leads to cuda toolkit right away. I would try not to make these checks in packages other than the cuda toolkit because cuda also supports clang (and other compilers), and keeping the same logic up to date in many places is... is... something bad.

Reply:
I fully agree with the ugliness of the string comparison. I will look at altering the toolkit ebuild method to check whether the user has a gcc version in the list. Just comparing the gcc version against the highest supported is not sufficient as it will not work when the list of supported versions does not include the lower gcc versions, or has holes.

Even if the gcc version is correct when the toolkit is built, I think it should be checked in the ebuild because the build error occurs here if it has been since changed using gcc-config before compiling opensubdiv, or by altered per package using package.env. The warning often points to cuda toolkit, but I think it increases usability to point out possible gcc version incompatibility as the cause. I acknowledge that this will not catch all cases, for example blender requires the appropriate gcc version to be correct at compile time as well as at runtime when it calls nvcc directly.

There are several packages that already do checks against cuda versions. For example chromium uses string comparison similar to the toolkit and tensorflow uses the method I have originally used here (likely with the same bug).

I also don't want to increase the maintenance burden, but given the functions it calls should be reasonably stable, the check should likely continue to work correctly once implemented. There are no hardcoded version numbers to update each revision. Perhaps adding this to the cuda eclass to allow reuse between ebuilds would be appropriate so that it can be maintained in one place only.